contracts/validator-manager/PoSValidatorManager.sol

```diff
@@ -486,46 +424,63 @@ abstract contract PoSValidatorManager is
+                    _getUptime(validationID, messageIndex);
             }
-            uint64 newValidatorWeight = validator.weight - delegator.weight;
-            (delegator.endingNonce,) = _setValidatorWeight(validationID, newValidatorWeight);
 
-            delegationEndTime = uint64(block.timestamp);
+            ($._delegatorStakes[delegationID].endingNonce,) =
+                _setValidatorWeight(validationID, validator.weight - delegator.weight);
+            $._redeemableDelegatorRewards[delegationID] =
+                _calculateDelegationReward(delegator, validator);
+        } else if (validator.status == ValidatorStatus.Completed) {
+            $._redeemableDelegatorRewards[delegationID] =
+                _calculateDelegationReward(delegator, validator);
+            return _completeEndDelegation(delegationID);
         } else {
-            // If the validation period has already ended, we have saved the uptime.
-            // Further, it is impossible to retrieve an uptime proof for an already ended validation,
-            // so there's no need to check any uptime proof provided in this function call.
-            validatorUptimeSeconds = $._completedValidationUptimeSeconds[validationID];
+            revert InvalidValidatorStatus();
+        }
 
-            delegator.endingNonce = validator.messageNonce;
+        emit DelegatorRemovalInitialized({delegationID: delegationID, validationID: validationID});
+    }
+
+    /// @dev Calculates the reward owed to the delegator based on the state of the delegator and its corresponding validator.
+    function _calculateDelegationReward(
+        Delegator memory delegator,
+        Validator memory validator
+    ) private view returns (uint256) {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        uint64 delegationEndTime;
+        if (
+            validator.status == ValidatorStatus.PendingRemoved
+                || validator.status == ValidatorStatus.Completed
+        ) {
             delegationEndTime = validator.endedAt;
+        } else if (validator.status == ValidatorStatus.Active) {
+            delegationEndTime = uint64(block.timestamp);
+        } else {
+            revert InvalidValidatorStatus();
         }
 
-        $._redeemableDelegatorRewards[delegationID] = $._rewardCalculator.calculateReward({
+        // Only give rewards in the case that the delegation started before the validator exited.
+        if (delegationEndTime <= delegator.startedAt) {
+            return 0;
+        }
+
+        return $._rewardCalculator.calculateReward({
             stakeAmount: weightToValue(delegator.weight),
             validatorStartTime: validator.startedAt,
             stakingStartTime: delegator.startedAt,
             stakingEndTime: delegationEndTime,
-            uptimeSeconds: validatorUptimeSeconds,
+            uptimeSeconds: $._completedValidationUptimeSeconds[delegator.validationID],
```

Comment:
It looks like "$._completedValidationUptimeSeconds[delegator.validationID]" is only updated in initializeEndDelegation if includeUptimeProof is true. Am I correct to assume that the delegation reward should be 0 if an uptime proof was never provided for this validationID?

Reply:
So this has changed a bit with Cam's PR [here](https://github.com/ava-labs/teleporter/pull/561). We cache the uptime proof provided, so this could have a value whether the delegator provides an uptime proof or not.